Shared/Samples/Download preplanned map area/DownloadPreplannedMapAreaView.Model.swift

```diff
@@ -0,0 +1,311 @@
+                    if model.isDownloading {
+                        group.addTask {
+                            await model.cancelDownloading()
+                        }
+                    }
+                }
+            }
+        }
+        
+        // Removes all downloaded maps.
+        func removeDownloadedMaps() {
+            // Sets the current map to the online web map.
+            selectedMap = .onlineWebMap
+            
+            // Update state.
+            canRemoveDownloadedMaps = false
+            
+            Task {
+                // Cancel all current download jobs.
+                await cancelAllJobs()
+                
+                // Remove each download.
+                allOfflineMapModels.forEach { $0.removeDownloadedContent() }
+            }
+        }
+    }
+}
+
+extension DownloadPreplannedMapAreaView.Model {
+    /// A type that specifies the currently selected map.
+    enum SelectedMap: Hashable {
+        /// The online version of the map.
+        case onlineWebMap
+        /// One of the offline maps.
+        case offlineMap(OfflineMapModel)
+    }
+}
+
+/// An object that encapsulates state about an offline map.
+class OfflineMapModel: ObservableObject, Identifiable {
+    /// The preplanned map area.
+    let preplannedMapArea: PreplannedMapArea
+    /// The task to use to take the area offline.
+    let offlineMapTask: OfflineMapTask
+    /// The directory where the mmpk will be stored.
+    let mmpkDirectory: URL
+    /// The currently running download job.
+    @Published private(set) var job: DownloadPreplannedOfflineMapJob?
+    /// The result of the download job.
+    @Published private(set) var result: Result<MobileMapPackage, Error>?
+    
+    init(preplannedMapArea: PreplannedMapArea, offlineMapTask: OfflineMapTask, temporaryDirectory: URL) {
+        self.preplannedMapArea = preplannedMapArea
+        self.offlineMapTask = offlineMapTask
+        self.mmpkDirectory = temporaryDirectory
+            .appendingPathComponent(preplannedMapArea.portalItem.id.rawValue)
+            .appendingPathExtension("mmpk")
+    }
+    
+    deinit {
+        Task { [job] in
+            // Cancel any outstanding job.
+            await job?.cancel()
+        }
+    }
+}
+
+extension OfflineMapModel: Hashable {
+    static func == (lhs: OfflineMapModel, rhs: OfflineMapModel) -> Bool {
+        lhs === rhs
+    }
+    
+    func hash(into hasher: inout Hasher) {
+        hasher.combine(ObjectIdentifier(self))
+    }
+}
+
+extension OfflineMapModel {
+    /// A Boolean value indicating whether the map is being taken offline.
+    var isDownloading: Bool {
+        job != nil
+    }
+}
+
+@MainActor
+private extension OfflineMapModel {
+    /// Downloads the given preplanned map area.
+    /// - Parameter preplannedMapArea: The preplanned map area to be downloaded.
+    /// - Precondition: `canDownload`
+    func download() async {
+        precondition(canDownload)
```

Comment:
It seems this precondition will never be hit, because on line 99 the download method is only executed when the offline map model can download. Is it here to solve some racing condition or timing issue?

Reply:
Correct it should never be hit. But if the code is changed, then it will let the developer know that something is wrong right away.